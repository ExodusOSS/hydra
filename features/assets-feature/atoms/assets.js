import { createInMemoryAtom } from '@exodus/atoms'

const assetsAtomDefinition = {
  id: 'assetsAtom',
  type: 'atom',
  factory: createInMemoryAtom,
  dependencies: [],
  public: true,
}

export default assetsAtomDefinition
