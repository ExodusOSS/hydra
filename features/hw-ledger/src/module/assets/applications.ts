const Bitcoin = {
  applicationName: 'Bitcoin',
  supportedVersions: '^2.1.2',
  primaryAssetName: 'bitcoin',
}

const Solana = {
  applicationName: 'Solana',
  supportedVersions: '^1.4.0',
  primaryAssetName: 'solana',
}

const Ethereum = {
  applicationName: 'Ethereum',
  supportedVersions: '^1.10.4',
  primaryAssetName: 'ethereum',
}

const Polygon = {
  applicationName: 'Polygon',
  supportedVersions: '^1.10.0',
  primaryAssetName: 'matic',
}

const Tron = {
  applicationName: 'Tron',
  supportedVersions: '^0.7.0',
  primaryAssetName: 'tronmainnet',
}

const applications = {
  Bitcoin,
  Solana,
  Ethereum,
  Polygon,
  Tron,
}

export default applications
