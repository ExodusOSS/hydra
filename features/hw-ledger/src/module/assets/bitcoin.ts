import bs58 from 'bs58'
import type { WalletPolicy } from '@exodus/ledger-bitcoin'
import { AppClient, PsbtV2 } from '@exodus/ledger-bitcoin'
import { BIP322 } from '@exodus/bip322-js'

import type Transport from '@ledgerhq/hw-transport'
import type { Atom } from '@exodus/atoms'

import type {
  Bytes,
  GetAddressParams,
  GetPublicKeyParams,
  GetXPubParams,
  HardwareWalletAssetHandler,
  SignMessageParams,
  SignTransactionParams,
  Signatures,
} from '@exodus/hw-common'
import { splitBip44DerivationPath } from './utils'
import assert from 'minimalistic-assert'
import {
  assertPsbtOnlyHasAllowedDerivationPaths,
  tagPsbtWithDerivationPaths,
  getWalletPolicy,
  getMultisigWalletPolicy,
  derivationPathsToWalletPolicy,
  addDummyInputs,
} from './utils/psbt'
import { buildBip322PSBT } from './utils/bip322'
import applications from './applications'

export interface BitcoinMessage {
  rawMessage?: Bytes

  bip322Message?: {
    address: string
    message: Bytes
  }

  bip137Message?: {
    address: string
    message: Bytes
  }
}

async function createHandler(
  transport: Transport,
  walletPolicyAtom?: Atom<Record<string, Buffer>>
): Promise<HardwareWalletAssetHandler<BitcoinMessage>> {
  const app = new AppClient(transport)
  const getAddress = async (params: GetAddressParams) => {
    const [purpose, coin, account, change, address] = splitBip44DerivationPath(
      params.derivationPath
    )
    const fpr = await app.getMasterFingerprint()
    const xpub = await app.getExtendedPubkey(`m/${purpose}/${coin}/${account}`)
    let walletPolicy
    let hmac = null
    if (params.multisigData) {
      walletPolicy = getMultisigWalletPolicy(fpr, xpub, purpose, coin, account, params.multisigData)
      hmac = await registerWalletPolicy(walletPolicy)
    } else {
      walletPolicy = getWalletPolicy(fpr, xpub, purpose, coin, account)
    }

    return app.getWalletAddress(
      walletPolicy,
      hmac,
      Number(change),
      Number(address),
      params.displayOnDevice ?? false // show address on the wallet's screen
    )
  }

  const getXPub = async (params: GetXPubParams) => {
    return app.getExtendedPubkey(params.derivationPath)
  }

  const getPublicKey = async (params: GetPublicKeyParams) => {
    return bs58.decode(await app.getExtendedPubkey(params.derivationPath)).slice(45, 45 + 33)
  }

  const signTransaction = async (params: SignTransactionParams) => {
    assert(Buffer.isBuffer(params.signableTransaction), 'signableTransaction must be PSBT buffer')

    const psbtBuffer = params.signableTransaction
    const psbt = new PsbtV2()
    psbt.deserialize(psbtBuffer)

    const fpr = await app.getMasterFingerprint()

    // Tag all of our own inputs and outputs with their corresponding
    // derivation information so the ledger knows which ones are ours.
    const derivationPathToPublicKeyMap = Object.create(null)
    const addressToDerivationPathMap = Object.create(null)
    for (const derivationPath of params.derivationPaths) {
      if (params.multisigData) {
        const address = await getAddress({
          assetName: 'bitcoin',
          derivationPath,
          multisigData: params.multisigData,
        })
        addressToDerivationPathMap[address] = derivationPath
      }

      const publicKey = await getPublicKey({
        assetName: 'bitcoin',
        derivationPath,
      })
      derivationPathToPublicKeyMap[derivationPath] = publicKey
    }

    tagPsbtWithDerivationPaths(psbt, fpr, derivationPathToPublicKeyMap, addressToDerivationPathMap)

    addDummyInputs(psbt)

    // Sanity check that the PSBT doesn't try to sign any
    // derivation paths that were not allowed listed.
    assertPsbtOnlyHasAllowedDerivationPaths(psbt, fpr, params.derivationPaths)

    const signatures: any[] = []
    const walletPolicies = await derivationPathsToWalletPolicy(
      fpr,
      app,
      params.derivationPaths,
      params.multisigData
    )

    for (const walletPolicy of walletPolicies) {
      const hmac = params.multisigData ? await registerWalletPolicy(walletPolicy) : null
      const _signatures = await app.signPsbt(psbt.serialize(), walletPolicy, hmac)
      signatures.push(..._signatures)
    }

    const result: Signatures = signatures.map(([inputIndex, partialSignature]) => ({
      inputIndex,
      publicKey: partialSignature.pubkey,
      signature: partialSignature.signature,
      tapleafHash: partialSignature.tapleafHash,
    }))

    return result
  }

  const signMessage = async (params: SignMessageParams<BitcoinMessage>) => {
    if (params.message.rawMessage) {
      assert(Buffer.isBuffer(params.message.rawMessage), 'rawMessage must be buffer')
      const signedMessageHex = await app.signMessage(
        params.message.rawMessage,
        params.derivationPath
      )
      return Buffer.from(signedMessageHex, 'base64')
    }

    if (params.message.bip322Message) {
      const { message, address } = params.message.bip322Message
      const { derivationPath } = params

      assert(Buffer.isBuffer(message), 'bip322Message.message must be buffer')
      assert(typeof address === 'string', 'bip322Message.address must be string')

      const [purpose, coin, account] = splitBip44DerivationPath(derivationPath)
      const fpr = await app.getMasterFingerprint()
      const xpub = await app.getExtendedPubkey(`m/${purpose}/${coin}/${account}`)
      const walletPolicy = getWalletPolicy(fpr, xpub, purpose, coin, account)

      const psbtBip322 = buildBip322PSBT(message.toString(), address)

      const psbt = new PsbtV2()
      psbt.deserialize(psbtBip322.toBuffer())

      // Tag our own inputs and outputs with their corresponding
      // derivation information so the ledger knows which ones are ours.
      const derivationPathToPublicKeyMap = {
        [derivationPath]: await getPublicKey({
          assetName: 'bitcoin',
          derivationPath,
        }),
      }
      tagPsbtWithDerivationPaths(psbt, fpr, derivationPathToPublicKeyMap)

      const signatures = await app.signPsbt(psbt.serialize(), walletPolicy, null)
      const signature = signatures[0]![1]

      const isTaprootSig = signature.pubkey.length === 32
      if (isTaprootSig) {
        psbtBip322.updateInput(0, {
          tapKeySig: signature.signature,
        })
      } else {
        psbtBip322.updateInput(0, {
          partialSig: [signature],
        })
      }

      psbtBip322.finalizeAllInputs()

      const encodedWitness = BIP322.encodeWitness(psbtBip322 as any)
      return Buffer.from(encodedWitness, 'base64')
    }

    throw new Error('rawMessage or bip322Message must be defined')
  }

  const registerWalletPolicy = async (walletPolicy: WalletPolicy): Promise<Buffer> => {
    const walletPolicies = (await walletPolicyAtom?.get()) ?? {}
    const id = walletPolicy.getId().toString('hex')
    let hmac = walletPolicies[id]
    if (!hmac) {
      const [, h] = await app.registerWallet(walletPolicy)
      hmac = h
      const updatedPolicies = { ...walletPolicies, [id]: hmac }
      await walletPolicyAtom?.set(updatedPolicies)
    }

    return hmac
  }

  return {
    getAddress,
    getPublicKey,
    getXPub,
    signTransaction,
    signMessage,
  }
}

const metadata = {
  applications: [applications.Bitcoin],
  handler: createHandler,
}

export default metadata
