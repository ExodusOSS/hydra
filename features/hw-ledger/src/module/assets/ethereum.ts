import * as secp256k1 from '@exodus/crypto/secp256k1'
import * as ethers from '@exodus/ethersproject-transactions'
import AppEth, { ledgerService } from '@exodus/ledgerhq-hw-app-eth'
import type { TypedMessage, MessageTypes } from '@metamask/eth-sig-util'
import { TypedDataUtils, SignTypedDataVersion } from '@metamask/eth-sig-util'
import assert from 'minimalistic-assert'

import type Transport from '@ledgerhq/hw-transport'
import type {
  GetAddressParams,
  GetPublicKeyParams,
  GetXPubParams,
  HardwareWalletAssetHandler,
  SignMessageParams,
  Message,
  SignTransactionParams,
} from '@exodus/hw-common'
import type { EIP712Message } from '@exodus/ledgerhq-hw-app-eth/lib/modules/EIP712/EIP712.types'

import { buildFauxXPUB } from './utils'
import { RequiresBlindSigningError } from '../errors'
import applications from './applications'

function hashEIP712Message(typedData: unknown) {
  const version = SignTypedDataVersion.V4

  const { sanitizeData, hashStruct } = TypedDataUtils
  const sanitizedData = sanitizeData(<TypedMessage<MessageTypes>>typedData)
  const domainSeparator = hashStruct(
    'EIP712Domain',
    sanitizedData.domain,
    sanitizedData.types,
    version
  )
  let hashStructMessage = Buffer.alloc(0)
  if (sanitizedData.primaryType !== 'EIP712Domain') {
    hashStructMessage = hashStruct(
      '' + sanitizedData.primaryType,
      sanitizedData.message,
      sanitizedData.types,
      version
    )
  }

  return {
    domainSeparator,
    hashStructMessage,
  }
}

function isEIP712Message(message: Record<string, unknown>): message is EIP712Message {
  return (
    typeof message === 'object' &&
    'types' in message &&
    'primaryType' in message &&
    'domain' in message &&
    'message' in message
  )
}

async function createHandler(transport: Transport): Promise<HardwareWalletAssetHandler> {
  const app = new AppEth(transport)

  return {
    getAddress: async (params: GetAddressParams) => {
      const { address } = await app.getAddress(params.derivationPath.replace('m/', ''))
      return address
    },

    getXPub: async (params: GetXPubParams) => {
      const { publicKey, chainCode } = await app.getAddress(
        params.derivationPath.replace('m/', ''),
        false,
        true
      )

      // Compress the public key in order to stay consistent with our codebase
      const compressedPublicKey = secp256k1.publicKeyConvert({
        publicKey: Buffer.from(publicKey, 'hex'),
        compressed: true,
        format: 'buffer',
      })

      return buildFauxXPUB(compressedPublicKey, Buffer.from(<string>chainCode, 'hex'))
    },

    getPublicKey: async (params: GetPublicKeyParams) => {
      // Ledger gives us an uncompressed public key in hex format
      const { publicKey } = await app.getAddress(params.derivationPath.replace('m/', ''))

      // Compress the public key in order to stay consistent with our codebase
      return secp256k1.publicKeyConvert({
        publicKey: Buffer.from(publicKey, 'hex'),
        compressed: true,
        format: 'buffer',
      })
    },

    signTransaction: async (params: SignTransactionParams) => {
      assert(
        Array.isArray(params.derivationPaths) && params.derivationPaths.length === 1,
        'derivationsPath array must be one element long'
      )
      assert(Buffer.isBuffer(params.signableTransaction), 'signableTransaction must be Buffer')

      const deserializedTransaction = ethers.parse(params.signableTransaction)

      /** TODO: retrieve from meta or bubble this up to the asset library "signHardware" */
      switch (params.assetName) {
        case 'ethereum': {
          deserializedTransaction.chainId = 1

          break
        }

        case 'matic': {
          deserializedTransaction.chainId = 137

          break
        }

        case 'basemainnet': {
          deserializedTransaction.chainId = 8453

          break
        }
        // No default
      }

      /** We need to delete the empty signature because otherwise ethers.js gets mad */
      delete deserializedTransaction.r
      delete deserializedTransaction.s
      delete deserializedTransaction.v

      // Serialize and cut off prefix '0x'
      const serializedTransactionHex = ethers.serialize(deserializedTransaction).slice(2)

      const signTransaction = async ({
        derivationPath,
        serializedTransactionHex,
        withNFT = true,
      }: {
        derivationPath: string
        serializedTransactionHex: string
        withNFT: boolean
      }) => {
        const resolution = await ledgerService.resolveTransaction(
          serializedTransactionHex,
          app.loadConfig,
          {
            externalPlugins: true,
            erc20: true,
            nft: withNFT,
          }
        )

        return app.signTransaction(derivationPath, serializedTransactionHex, resolution)
      }

      let response = null

      try {
        response = await signTransaction({
          derivationPath: params.derivationPaths[0],
          serializedTransactionHex,
          withNFT: true,
        })
      } catch (err: any) {
        if (err.name === 'EthAppNftNotSupported') {
          // Attempt to recover from this error by signing the transaction blindly
          // Alternatively we just need to check if device is Nano S before trying this
          // Note: sending NFTs on Nano S requires blind signing to be enabled
          response = await signTransaction({
            derivationPath: params.derivationPaths[0],
            serializedTransactionHex,
            withNFT: false,
          })
        } else if (err.name === 'EthAppPleaseEnableContractData') {
          throw new RequiresBlindSigningError()
        } else {
          throw err
        }
      }

      // Decomposed signature
      const { v, r, s } = response

      return [
        {
          signature: Buffer.concat([
            Buffer.from(v, 'hex'),
            Buffer.from(r, 'hex'),
            Buffer.from(s, 'hex'),
          ]),
        },
      ]
    },

    signMessage: async (params: SignMessageParams) => {
      const message = params.message as Message & {
        EIP712Message?: EIP712Message
      }

      let signingPromise

      if (message.rawMessage) {
        assert(Buffer.isBuffer(params.message.rawMessage), 'rawMessage must be a buffer')
        signingPromise = app.signPersonalMessage(
          params.derivationPath,
          Buffer.from(message.rawMessage).toString('hex')
        )
      } else if (message.EIP712Message) {
        assert(
          isEIP712Message(message.EIP712Message),
          'EIP712Message must be EIP712 conformant message'
        )
        signingPromise = app
          .signEIP712Message(params.derivationPath, message.EIP712Message)
          .catch((err: Error) => {
            // Signing EIP712 messages is not supported on the Nano S
            // Fallback to blind signing the message
            if (err.message.includes('INS_NOT_SUPPORTED')) {
              // Manually compute the domain & message hashes
              const { domainSeparator, hashStructMessage } = hashEIP712Message(
                message.EIP712Message
              )
              return app.signEIP712HashedMessage(
                params.derivationPath,
                domainSeparator.toString('hex'),
                hashStructMessage.toString('hex')
              )
            }

            throw err
          })
      } else {
        assert(false, 'rawMessage or EIP712Message must be specified')
      }

      const { r, s, v } = await signingPromise
      const vNorm = v.toString(16)
      const vHex = vNorm.length < 2 ? '0' + vNorm : vNorm
      return Buffer.from(`${r}${s}${vHex}`, 'hex')
    },
  }
}

const metadata = {
  applications: [applications.Ethereum],
  handler: createHandler,
}

export default metadata
