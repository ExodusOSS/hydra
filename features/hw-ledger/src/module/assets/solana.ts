/* eslint-disable @typescript-eslint/no-unused-vars */
import AppSolana from '@ledgerhq/hw-app-solana'
import assert from 'minimalistic-assert'
import bs58 from 'bs58'
import { XPubUnsupportedError } from '@exodus/hw-common'

import type Transport from '@ledgerhq/hw-transport'
import type {
  GetAddressParams,
  GetPublicKeyParams,
  GetXPubParams,
  HardwareWalletAssetHandler,
  SignMessageParams,
  SignTransactionParams,
} from '@exodus/hw-common'

import { RequiresBlindSigningError } from '../errors'
import applications from './applications'

async function createHandler(transport: Transport): Promise<HardwareWalletAssetHandler> {
  const app = new AppSolana(transport)

  const getAddress = async (params: GetAddressParams) => {
    const { address } = await app.getAddress(params.derivationPath.replace('m/', ''))
    return bs58.encode(address)
  }

  const getXPub = async (params: GetXPubParams) => {
    throw new XPubUnsupportedError()
  }

  const getPublicKey = async (params: GetPublicKeyParams) => {
    // Addresses in Solana are just Base58 encoded strings of 32 byte ed25519 public keys.
    const { address } = await app.getAddress(params.derivationPath.replace('m/', ''))
    return address
  }

  const signTransaction = async (params: SignTransactionParams) => {
    assert(
      Array.isArray(params.derivationPaths) && params.derivationPaths.length === 1,
      'derivationsPath array must be one element long'
    )

    let serializedTransaction
    if (Buffer.isBuffer(params.signableTransaction)) {
      serializedTransaction = params.signableTransaction
    } else if (params.signableTransaction instanceof Uint8Array) {
      serializedTransaction = Buffer.from(params.signableTransaction)
    } else if ((params.signableTransaction as any).serialize) {
      serializedTransaction = Buffer.from((params.signableTransaction as any).serialize())
    }

    if (!Buffer.isBuffer(serializedTransaction)) {
      throw new TypeError(`signableTransaction must be Buffer or serializable to a Buffer.`)
    }

    const { signature } = await app
      .signTransaction(params.derivationPaths[0].replace('m/', ''), serializedTransaction)
      .catch((err: any) => {
        if (err.message.includes('Try enabling blind signature in the app')) {
          throw new RequiresBlindSigningError()
        } else {
          throw err
        }
      })

    const { address: publicKey } = await app.getAddress(params.derivationPaths[0].replace('m/', ''))

    return [
      {
        publicKey,
        signature,
      },
    ]
  }

  const signMessage = async (params: SignMessageParams) => {
    assert(Buffer.isBuffer(params.message.rawMessage), 'rawMessage must be a buffer')
    const { signature } = await app.signOffchainMessage(
      params.derivationPath.replace('m/', ''),
      params.message.rawMessage
    )
    return signature
  }

  return {
    getAddress,
    getPublicKey,
    getXPub,
    signTransaction,
    signMessage,
  }
}

const metadata = {
  applications: [applications.Solana],
  handler: createHandler,
}

export default metadata
