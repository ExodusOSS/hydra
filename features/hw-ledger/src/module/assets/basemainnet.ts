import applications from './applications'
import ethereum from './ethereum'

const metadata = {
  applications: [applications.Ethereum],
  handler: ethereum.handler,
}

export default metadata
