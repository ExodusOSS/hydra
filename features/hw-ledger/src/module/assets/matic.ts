import applications from './applications'
import ethereum from './ethereum'

const metadata = {
  applications: [applications.Polygon, applications.Ethereum],
  handler: ethereum.handler,
}

export default metadata
