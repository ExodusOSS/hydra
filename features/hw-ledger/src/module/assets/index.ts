import bitcoin from './bitcoin'
import basemainnet from './basemainnet'
import ethereum from './ethereum'
import matic from './matic'
import solana from './solana'
import tronmainnet from './tronmainnet'
import applications from './applications'

import type Transport from '@ledgerhq/hw-transport'
import type { HardwareWalletAssetHandler } from '@exodus/hw-common'
import type { Atom } from '@exodus/atoms'

type HandlerFactory = (
  transport: Transport,
  walletPolicyAtom?: Atom<Record<string, Buffer>>
) => Promise<HardwareWalletAssetHandler>

type AssetMetadata = {
  applications: {
    applicationName: string
    primaryAssetName: string
    supportedVersions: string
  }[]
  handler: HandlerFactory
}

export const assetApplications: Record<string, AssetMetadata> = {
  basemainnet,
  bitcoin,
  bitcointestnet: bitcoin,
  bitcoinregtest: bitcoin,
  matic,
  solana,
  ethereum,
  ethereumsepolia: ethereum,
  tronmainnet,
}

/**
 * The baseAssetNames that are supported by our implementation.
 * e.g: ['bitcoin', 'ethereum', 'matic', 'solana']
 */
export const supportedBaseAssetNames = Object.keys(assetApplications)

/**
 * An object containing a map that maps the asset application name
 * to their respective baseAssetName.
 * e.g { "Bitcoin": ["bitcoin"], "Ethereum": ["ethereum", "matic"], }
 */
export const applicationNameToBaseAssetNames: Record<string, string[]> = Object.entries(
  assetApplications
).reduce((mapping, [assetName, metadata]) => {
  // Loop over all supported applications for a given asset name
  for (const { applicationName } of metadata.applications) {
    if (!Array.isArray(mapping[applicationName])) {
      // Create an empty array if it doesn't exist
      mapping[applicationName] = []
    }

    const isPrimaryAssetForApplication: boolean =
      (applications as any)[applicationName]?.primaryAssetName === assetName
    // Add the assetName to the mapping
    if (isPrimaryAssetForApplication) {
      // Push to front of array, since the first assetName in the list
      // is the primary asset to display
      mapping[applicationName].unshift(assetName)
    } else {
      // Push at the end of the array
      mapping[applicationName].push(assetName)
    }
  }

  return mapping
}, Object.create(null))
