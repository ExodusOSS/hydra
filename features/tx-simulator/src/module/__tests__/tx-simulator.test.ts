import { createSimulationServices } from '@exodus/web3-simulation'
import assets from '@exodus/assets-base'
import { deserializeTransaction } from '@exodus/web3-solana-utils'

import txSimulatorDefinition from '../'

jest.mock('@exodus/web3-simulation')

const assetsModule = {
  getAsset: (name: string) => assets[name],
}

const createTxSimulator = () => {
  return txSimulatorDefinition.factory({
    assetsModule,
    assetClientInterface: {},
    config: {
      apiUrl: 'https://simulation.a.exodus.io',
    },
    logger: {
      warn: () => {}, // noop
    },
  })
}

assets.solana.api = {
  getFeeData: () => Promise.resolve({ fee: assets.solana.currency.defaultUnit(5000) }),
}
assets.solana.baseAsset = assets.solana

describe('tx-simulator', () => {
  it('should simulate a solana transaction', async () => {
    console.log({ createSimulationServices })
    ;(createSimulationServices as jest.Mock).mockImplementation(() => ({
      simulateSolanaTransactions: () => ({
        advancedDetails: [
          {
            instructions: [
              {
                formattedValue: 'Comp..1111',
                name: 'Program Id',
                value: 'ComputeBudget111111111111111111111111111111',
              },
              {
                formattedValue: 'AsBcFQA=',
                name: 'Raw Data',
                value: 'AsBcFQA=',
              },
            ],
            title: 'Unknown',
          },
          {
            instructions: [
              {
                formattedValue: 'Comp..1111',
                name: 'Program Id',
                value: 'ComputeBudget111111111111111111111111111111',
              },
              {
                formattedValue: 'AwQXAQAAAAAA',
                name: 'Raw Data',
                value: 'AwQXAQAAAAAA',
              },
            ],
            title: 'Unknown',
          },
          {
            instructions: [
              {
                formattedValue: 'ATok..8knL',
                name: 'Program Id',
                value: 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL',
              },
            ],
            title: 'Create Token Account',
          },
          {
            instructions: [
              {
                formattedValue: '8cBx..8Yws',
                name: 'From Public Key',
                value: '8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws',
              },
              {
                formattedValue: 'BWNe..9CSR',
                name: 'To Public Key',
                value: 'BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR',
              },
              {
                formattedValue: '0.1048 SOL',
                name: 'Amount',
                value: '0.1048 SOL',
              },
            ],
            title: 'Transfer SOL',
          },
          {
            instructions: [
              {
                formattedValue: 'BWNe..9CSR',
                name: 'Public Key',
                value: 'BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR',
              },
            ],
            title: 'Create Sync Native Account',
          },
          {
            instructions: [
              {
                formattedValue: 'JUP6..TaV4',
                name: 'Program Id',
                value: 'JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4',
              },
              {
                formattedValue: 'wSCb..VQ==',
                name: 'Raw Data',
                value: 'wSCbM0HWnIEEAQAAABwAZAABSkQ/BgAAAADXG6/dFgAAAMgAVQ==',
              },
            ],
            title: 'Unknown',
          },
          {
            instructions: [
              {
                formattedValue: 'BWNe..9CSR',
                name: 'Source Public Key',
                value: 'BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR',
              },
              {
                formattedValue: '8cBx..8Yws',
                name: 'Destination Public Key',
                value: '8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws',
              },
              {
                formattedValue: '8cBx..8Yws',
                name: 'Owner Public Key',
                value: '8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws',
              },
            ],
            title: 'Close Account',
          },
        ],
        balanceChanges: {
          willApprove: [],
          willPayFee: [
            {
              balance: {
                type: 'NumberUnit',
                unit: 'SOL',
                unitType: 'SOL',
                value: '5000',
              },
            },
          ],
          willReceive: [
            {
              balance: {
                type: 'NumberUnit',
                unit: 'base',
                unitType: 'Bonk',
                value: '100934700000',
              },
            },
          ],
          willSend: [
            {
              balance: {
                type: 'NumberUnit',
                unit: 'base',
                unitType: 'SOL',
                value: '104809546',
              },
            },
          ],
        },
        baseAssetName: 'solana',
        displayDetails: {
          warnings: [],
        },
        transactions: [],
      }),
    }))

    const simulationResult = await createTxSimulator().simulate({
      assetName: 'solana',
      transactions: [
        deserializeTransaction(
          'AQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAQAIDnEFiG4DTZRCbmnlsIoI3aN48Z+uAD3YHPdxGxDNU+kyWO9nf7VjXmRzcktw4WtkBVQDTqR6HHs/zYiFPEFdMlRz5cjChx7fASyDZXpi4eeIlroBba7O6Ha71C+pOt3deJwbMmsJnEMjC0q1UfKO/vm8lB7xy8Qv2bts1M7ijF6u1JIwjXYeeEPjR4K57wzJvgWOZAZSwyw3treT+Ax5/Dnd9SAxdDGH/Fn4MZNi0fkAVLnSg74IDSJ337fzl3gWCQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwZGb+UhFzL/7K26csOb57yM5bvF9xJrLEObOkAAAAAEedVb8jHAbu50xW7OaBUH/bGy3qP0jlECsc2iVrwTjwbd9uHXZaGT2cvhRs7reawctIXtX1s3kTqM9YV+/wCpOriQP7c1yrHGfFmvSFft9hsK+DKlCnxZ4yGRng7IqbyMlyWPTiSJ8bs9ECkUjg2DC1oTmdr/EIQEjnvY2+n4WbQ/+if11/ZKdMCbHylYed5LCas238ndUUsyGqezjOXovAfFbmCtPT8Xc4LqxlSPuh/TLP2QygKz58+hhf3Oc5ha/PTzYK2uaCqy9gLVLXh6yn1xtnMYT0oGzv5G667rSQcHAAUCwFwVAAcACQMEFwEAAAAAAAsGAAMAEgYJAQEGAgADDAIAAABKRD8GAAAAAAkBAwERCBYJCgADAQQCEg0FCAwIERMPCgQBEA4JJcEgmzNB1pyBBAEAAAAcAGQAAUpEPwYAAAAA1xuv3RYAAADIAFUJAwMAAAEJAWhrKDhvF3v5paQOu9+BZi1gmEU3DJz8lRgcZwTCQBuZA/n7+gP2hvc='
        ),
      ],
      origin: 'https://jup.ag',
      senderAddress: '8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws',
    })

    expect(simulationResult).toMatchSnapshot()
  })
})
