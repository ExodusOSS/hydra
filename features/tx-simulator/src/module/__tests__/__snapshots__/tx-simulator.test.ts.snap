// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tx-simulator should simulate a solana transaction 1`] = `
{
  "advancedDetails": [
    {
      "instructions": [
        {
          "formattedValue": "Comp..1111",
          "name": "Program Id",
          "value": "ComputeBudget111111111111111111111111111111",
        },
        {
          "formattedValue": "AsBcFQA=",
          "name": "Raw Data",
          "value": "AsBcFQA=",
        },
      ],
      "title": "Unknown",
    },
    {
      "instructions": [
        {
          "formattedValue": "Comp..1111",
          "name": "Program Id",
          "value": "ComputeBudget111111111111111111111111111111",
        },
        {
          "formattedValue": "AwQXAQAAAAAA",
          "name": "Raw Data",
          "value": "AwQXAQAAAAAA",
        },
      ],
      "title": "Unknown",
    },
    {
      "instructions": [
        {
          "formattedValue": "ATok..8knL",
          "name": "Program Id",
          "value": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
        },
      ],
      "title": "Create Token Account",
    },
    {
      "instructions": [
        {
          "formattedValue": "8cBx..8Yws",
          "name": "From Public Key",
          "value": "8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws",
        },
        {
          "formattedValue": "BWNe..9CSR",
          "name": "To Public Key",
          "value": "BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR",
        },
        {
          "formattedValue": "0.1048 SOL",
          "name": "Amount",
          "value": "0.1048 SOL",
        },
      ],
      "title": "Transfer SOL",
    },
    {
      "instructions": [
        {
          "formattedValue": "BWNe..9CSR",
          "name": "Public Key",
          "value": "BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR",
        },
      ],
      "title": "Create Sync Native Account",
    },
    {
      "instructions": [
        {
          "formattedValue": "JUP6..TaV4",
          "name": "Program Id",
          "value": "JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4",
        },
        {
          "formattedValue": "wSCb..VQ==",
          "name": "Raw Data",
          "value": "wSCbM0HWnIEEAQAAABwAZAABSkQ/BgAAAADXG6/dFgAAAMgAVQ==",
        },
      ],
      "title": "Unknown",
    },
    {
      "instructions": [
        {
          "formattedValue": "BWNe..9CSR",
          "name": "Source Public Key",
          "value": "BWNevwu92XRSD3c9DBhZrq9JFMUGcN9nm1dmKxY19CSR",
        },
        {
          "formattedValue": "8cBx..8Yws",
          "name": "Destination Public Key",
          "value": "8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws",
        },
        {
          "formattedValue": "8cBx..8Yws",
          "name": "Owner Public Key",
          "value": "8cBxsCL2q2jMXqHweMaHmP7v2RbB4WG5buTXmjXa8Yws",
        },
      ],
      "title": "Close Account",
    },
  ],
  "balanceChanges": {
    "willApprove": [],
    "willPayFee": [
      {
        "balance": {
          "type": "NumberUnit",
          "unit": "SOL",
          "unitType": "SOL",
          "value": "5000",
        },
      },
    ],
    "willReceive": [
      {
        "balance": {
          "type": "NumberUnit",
          "unit": "base",
          "unitType": "Bonk",
          "value": "100934700000",
        },
      },
    ],
    "willSend": [
      {
        "balance": {
          "type": "NumberUnit",
          "unit": "base",
          "unitType": "SOL",
          "value": "104809546",
        },
      },
    ],
  },
  "baseAssetName": "solana",
  "displayDetails": {
    "warnings": [],
  },
  "transactions": [],
}
`;
